### DISCUSSION TITLE
How results of watch() call are generated?

### DISCUSSION THREAD

**Discussion by givip** [simolus3/drift#1025] [Q&A]
I've noticed that watch() generates empty lists of results sometimes? When this happens?

Or it triggers multiple times with same result. 

What is the exact algorithm of watch(), I want to understand how results are generated.

**Comment by simolus3** [comment]
> Or it triggers multiple times with same result.

This can happen. The stream is invalidated whenever any referenced table is updated. That doesn't necessarily mean that the query will emit different results though.

> I've noticed that watch() generates empty lists of results sometimes?

That's pretty weird, are you sure that there's nothing wrong with your query? If you have reproducible code that doesn't behave as expected, please open an issue.

> What is the exact algorithm of watch(), I want to understand how results are generated.

You can find the implementation [here](https://github.com/simolus3/moor/blob/master/moor/lib/src/runtime/executor/stream_queries.dart). In particular, the `StreamQueryStore` and the `QueryStream` classes are relevant here.
